### DISCUSSION TITLE
Fetching posts by the authors I follow

### DISCUSSION THREAD

**Discussion by shajidhasan** [pocketbase/pocketbase#5314] [Q&A] [answered]
So, I am basically building a Twitter clone, and I'm trying to fetch the posts created by the users the current user follows.

## Attempt 1

I tried having the following collection structures:
* `users`
	* `id`
	* `followers` (multiple `relation` to `users`)
* `posts`
	* `author` (single `relation` to `users`)
	* `content`

This allows me to fetch `posts` with a query like this `"author.followers.id ?= 'currentUserId'"` from the posts collection. `currentUserId` being here the logged in user. This works pretty well.

### The issue
The drawback is that, every time I view/list users, or anything else that has the `author`/`user` relation expanded, it'll come with a huge array of followers which I will not need. One solution I found, is that I can use the `fields` option to limit which fields the API returns, but I haven't been able to make that work in an expanded relation.

## Attempt 2

I tried having the following collection structures:
* `users`
	* `id`
* `follows`
	* `follower` (single `relation` to `users`)
	* `followed` (single `relation` to `users`)	
* `posts`
	* `author` (single `relation` to `users`)
	* `content`

This structure has allowed me to use this filter on the `posts`: `"@collection.follows.follower ?= 'currentUserId' && author = @collection.follows.followed"`.

### The issue
I soon learned that the `@collection` queries can only be done by admins.

I also tried to utilize the view collections and I had a few ideas. But unfortunately, my SQL knowledge is pretty limited, and I could not get anywhere.

So, any idea how should I achieve this in an efficient way? Thank you so much.

**Comment by ganigeorgiev** [comment, marked as answer]
You've mentioned that you've tried several solution but nowhere in the above shows what actually you've tried and what were the actual issues you've encountered.

I'm not entirely sure that I understand your described collections structure because the used names don't seem to correspond to your described fields in the attempted API rule.

From where the `currentUserId` string is expected to come from? Is it a value you want to send with the request?

What actually is the desired behavior? Are you trying to restrict access or just to filter the posts (aka. all users can read all posts and you just want to fetch the posts matching your criteria)?

If we assume the latter and the following collections structure:

- `users`
   - `id`

- `followers`
  - `user` (single `relation` to `users`)
  - `follow` (single `relation` to `users`)
  - add combined UNIQUE index for the 2 fields to prevent duplicated entries

- `posts`
   - `author` (single `relation` to `users`; an index on the field could help with the lookups)

Then if you just want to fetch with the JS SDK list of the posts created from other users that the current one is following, you could do something like:

```js
const currentUserId = pb.authStore.model.id;

await pb.collection("posts").getList(1, 30, {
  // see https://pocketbase.io/docs/working-with-relations/#back-relations
  filter: `author.followers_via_follow.user = "${currentUserId}"`,
})
```

If that doesn't help please be more specific about your current collections configuration (_you can also export it as json from the Admin UI > Settings > Export collections_) and the desired behavior.

**Reply by shajidhasan** [reply]
That worked wonders! Thank you so much! And also, many thanks for developing and maintaining such an amazing project. And painstakingly helping everyone too.

And I'm extremely sorry for not being clear enough. I have edited the post, should it prove useful to someone in the future.
